### PR TITLE
[ML] Ensure class is represented when its cardinality is low

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
In DF analytics classification, it is possible to use no samples
of a class if its cardinality is too low.

This commit fixes this by ensuring the target sample count can never be zero.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java [hunk 1]
```diff
@@ -57,9 +57,13 @@ public void process(String[] row, Runnable incrementTrainingDocs, Runnable incre
             throw new IllegalStateException("Unknown class [" + classValue + "]; expected one of " + classSamples.keySet());
         }
 
+        // We use ensure the target sample count is at least 1 as if the cardinality
+        // is too low we might get a target of zero and, thus, no samples of the whole class
+        double targetSampleCount = Math.max(1.0, samplingRatio * sample.cardinality);
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: Is it guaranteed that `sample.cardinality` is at least `1`? If it was `0`, then `targetSampleCount` would be `1` but maybe that's not what we'd wanted?

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: Wouldn't it be easier to just take the first value found?

`boolean isTraining = random.nextDouble() <= p || sample.observed == 0;`

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: @przemekwitek If cardinality was zero then we'd never encounter a doc with that class.

However, @benwtrent 's suggestion also works. I'll change it to that as it seems simpler indeed.

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: Actually, as soon as I made that change the unit test we have for checking we uniformly pick samples failed. As this solution means we always pick the first value we see, we introduce a bias towards that. I'll stick to my original solution as it prevents that bias.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: @dimitris-athanasiou I thought it was not possible to even run a classification job if cardinality was `< 2`. The test talks about a cardinality of `1`.

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: It makes no difference from the perspective of the code but I'll update to use cardinality of `2` for the sake of reflecting actual conditions.

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: Also, note these are different cardinalities we're talking about. The requirement of `2` refers to "there should be at least 2 classes". Whereas the addressed problem arises when the cardinality of the docs belonging to a certain class is low.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: > Whereas the addressed problem arises when the cardinality of the docs belonging to a certain class is low.

AH! This is confusing 🤔 . The solution as posed makes sense.

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:62]: I have pushed a commit that should make the unit test clearer and more realistic.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java [hunk 2]
```diff
@@ -57,9 +57,13 @@ public void process(String[] row, Runnable incrementTrainingDocs, Runnable incre
             throw new IllegalStateException("Unknown class [" + classValue + "]; expected one of " + classSamples.keySet());
         }
 
+        // We use ensure the target sample count is at least 1 as if the cardinality
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java:60]: "We use ensure"
Something is wrong with this sentence.

**Review by przemekwitek — APPROVED** [review]
LGTM

Just one small comment regarding... comment.

**Pull request merged**